File: WeeklyCode/week5/Path Sum/chenhao.java
public class Solution {
    
    private boolean dfs(TreeNode cur,int sum,int curSum){
        curSum+=cur.val;
        if(cur.left==null && cur.right==null){
            return sum==curSum;
        }
        return ( cur.left!=null && dfs(cur.left,sum,curSum) ) ||
            ( cur.right!=null && dfs(cur.right,sum,curSum) );
    }
    
    public boolean hasPathSum(TreeNode root, int sum) {
        if(root==null){
            return false;
        }
        return dfs(root,sum,0);
    }
}

File: WeeklyCode/week5/Unique Binary Search Trees II/chenhao.py

class Solution:
    
    cache={}
    
    def cloneTree(self,root):
        if root==None:
            return None
        clone=TreeNode(root.val)
        clone.left=self.cloneTree(root.left)
        clone.right=self.cloneTree(root.right)
        return clone
    
    def generateTreesStructure(self,n):
        if n==0:
            return [None]
        if n==1:
            return [TreeNode(0)]
        if self.cache.get(n)!=None:
            return self.cache.get(n)
        res=[]
        for nLeft in range(0,n):
            nRight=n-1-nLeft
            for left in self.generateTreesStructure(nLeft):
                for right in self.generateTreesStructure(nRight):
                    root=TreeNode(0)
                    root.left=self.cloneTree(left)
                    root.right=self.cloneTree(right)
                    res.append(root)
        self.cache[n]=res
        return res
    
    def assignValues(self,cur):
        if cur==None:
            return
        self.assignValues(cur.left)
        cur.val=self.curVal
        self.curVal+=1
        self.assignValues(cur.right)
    
    # @return a list of tree node
    def generateTrees(self, n):
        trees=self.generateTreesStructure(n)
        for root in trees:
            self.curVal=1
            self.assignValues(root)
        return trees